File: backend/app/models/notes/note.py
# Copyright (c) 2026 Valaris Studio
# SPDX-License-Identifier: AGPL-3.0-or-later

import uuid

from sqlalchemy import JSON, Boolean, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, UUIDMixin


class Note(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "notes"

    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workspaces.id", ondelete="CASCADE"), index=True
    )
    board_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("boards.id", ondelete="CASCADE"), nullable=True, index=True
    )
    card_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("cards.id", ondelete="SET NULL"), nullable=True, index=True
    )
    title: Mapped[str] = mapped_column(String(500))
    content: Mapped[str] = mapped_column(Text, default="")
    # Plain-text projection of `content`, maintained by NoteService on every
    # write. Search ILIKEs this instead of `content` — the raw column is
    # ProseMirror JSON, so a LIKE over it matches node names ("paragraph",
    # "heading") on every note. Also serves the list preview without parsing
    # each document per row. NULLABLE: rows written by pre-migration-099 code
    # during a rolling deploy have no value yet, and readers fall back to
    # extracting from `content` (see services.notes.content_text.build_preview).
    content_text: Mapped[str | None] = mapped_column(Text, nullable=True, default=None)
    pinned: Mapped[bool] = mapped_column(Boolean, default=False)
    # See app.models.notes.kinds — kept as a free String (not Enum) so the set
    # of kinds is operator-extensible without a schema change.
    kind: Mapped[str] = mapped_column(
        String(32), nullable=False, server_default="user_note", default="user_note"
    )
    # Reviewer-only field. NULL on every other kind and on approving verdicts.
    # See app.models.notes.failure_class.ReviewFailureClass for the closed set
    # of values; the column is a plain string so old code (pre SWE-AF #5) keeps
    # working — NULL just means "no classification recorded".
    failure_class: Mapped[str | None] = mapped_column(
        String(32), nullable=True, default=None
    )
    # Reviewer-only field (SWE-AF #3). Structured severity-tiered findings.
    # Shape: list[{severity, message, file?, function?}] where severity is one
    # of BLOCKING / SHOULD_FIX / SUGGESTION (see app.models.notes.finding).
    # NULL on every other kind and on legacy verdicts pre SWE-AF #3 — the
    # derivation helper (services.notes.note.derive_approved) treats NULL as
    # "no structured findings, fall back to the legacy decision string".
    findings: Mapped[list | None] = mapped_column(JSON, nullable=True, default=None)
    # The execution that produced this note, for machine-generated notes
    # (review verdicts, system notes). Lets the UI link a note to its source
    # execution (the "link the mentioned element" rule). NULL for human notes
    # and pre-rollout rows. ondelete=SET NULL so reaping an execution doesn't
    # delete the verdict it wrote.
    source_execution_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("agent_executions.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id")
    )

File: frontend/src/types/note.ts
// Copyright (c) 2026 Valaris Studio
// SPDX-License-Identifier: AGPL-3.0-or-later

export interface NoteFinding {
  severity: "BLOCKING" | "SHOULD_FIX" | "SUGGESTION";
  message: string;
  file?: string | null;
  function?: string | null;
}

export interface Note {
  id: string;
  workspace_id: string;
  board_id: string | null;
  card_id: string | null;
  title: string;
  content: string;
  pinned: boolean;
  // See backend app/models/notes/kinds.py. Free string (operator-extensible);
  // defaults to "user_note". The frontend catalog (features/notes/lib/noteKinds)
  // maps known kinds to a label/tooltip and an agent-vs-human origin.
  kind: string;
  // Reviewer-only; NULL on every other kind and on approving verdicts.
  failure_class: string | null;
  // Reviewer-only structured findings; NULL otherwise.
  findings: NoteFinding[] | null;
  // The execution that produced this note, when it was machine-generated
  // (review verdict, system note). NULL for human notes or pre-rollout rows.
  // Lets the UI link a note straight to its source execution (house "link the
  // element" rule). See backend app/models/notes/note.py.
  source_execution_id: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

/**
 * A note as it appears in the LIST. The body is deliberately absent: a
 * workspace with a few hundred large notes ships megabytes of ProseMirror JSON
 * if the browse payload carries `content`. The server sends `preview` (plain
 * text, markup already stripped) for the card/row snippet; anything that needs
 * the real body reads it through `useNote` (the detail endpoint).
 */
export interface NoteSummary extends Omit<Note, "content"> {
  preview: string;
}

export interface NoteCreate {
  title: string;
  content?: string;
  pinned?: boolean;
  card_id?: string | null;
}

export interface NoteUpdate {
  title?: string;
  content?: string;
  pinned?: boolean;
  // null unlinks the note from its card.
  card_id?: string | null;
}
